### PR TITLE
[ISSUE #614] Add schema plugin and its api

### PR DESCRIPTION
**Pull request by jzhou59**
Fixes ISSUE #614 .

### Motivation

This pr adds a plugin for schema and adds the api for it.

### Documentation
This pr introduces a new feature schema client. 
The general schema design is presented in the repo, but the client is not documented yet.
I will update the schema documentation later.

### CODE REVIEW

File: eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaPluginFactory.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+     * Get {@code SchemaRegistry}
+     *
+     * @param schemaRegistryType the name of schema registry that configured in properties file
+     * @return schema registry plugin or null
+     */
+    public static SchemaRegistry getSchemaRegistry(String schemaRegistryType) {
+        checkNotNull(schemaRegistryType, "SchemaRegistryType cannot be null");
+        SchemaRegistry schemaRegistry = EventMeshExtensionFactory.getExtension(SchemaRegistry.class, schemaRegistryType);
+        return checkNotNull(schemaRegistry, "MetricsRegistryType: " + schemaRegistryType + " is not supported");
+    }
```
Comment (ruanwenjun) [inline eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaPluginFactory.java:39]:
```suggestion
        return checkNotNull(schemaRegistry, "SchemaRegistryType: " + schemaRegistryType + " is not supported");
```

File: eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaType.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+public enum SchemaType {
+    JSON("JSON"),
+    AVRO("AVRO"),
+    PROTOBUF("PROTOBUF");
+
+    private final String schemaTypeName;
+
+    SchemaType(String schemaTypeName) {
+        this.schemaTypeName = StringUtils.upperCase(StringUtils.deleteWhitespace(schemaTypeName));
+    }
```
Comment (ruanwenjun) [inline eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaType.java:31]: Use `this.schemaTypeName = schemaTypeName` is better.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]
You may need to add `gradle.properties`, this will be used when packaging.

BTY, I think your `SchemaRegistry` design is not great. I suggest the schema just need to considerate how to serialize and deserialize. It doesn't need to take care about how to register schema, this should be implemen at `RegisterCenter` plugin.

File: eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaPluginFactory.java (hunk 1 above)
Comment (jzhou59) [reply eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaPluginFactory.java]: fixed

File: eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaType.java (hunk 2 above)
Comment (jzhou59) [reply eventmesh-schema-plugin/eventmesh-schema-api/src/main/java/org/apache/eventmesh/schema/api/SchemaType.java]: Ok, I was trying to specify all schema types to uppercase.
But you are right, I have altered the schema type according to IETF(or official) abbreviation.

References:
- [JSON](https://datatracker.ietf.org/doc/html/rfc8259)
- [Avro](https://avro.apache.org/docs/current/spec.html)
- [Protobuf](https://datatracker.ietf.org/doc/html/draft-rfernando-protocol-buffers-00#section-1.1)

**Comment by jzhou59** [comment]
> You may need to add `gradle.properties`, this will be used when packaging.
> 
> BTY, I think your `SchemaRegistry` design is not great. I suggest the schema just need to considerate how to serialize and deserialize. It doesn't need to take care about how to register schema, this should be implemen at `RegisterCenter` plugin.

Sure, currently it is only ```api``` package and I will add ```gradle.properties``` in implementations.

The registering-related and retrieving-related procedures are indeed better in ```registry```-related plugin and I will modify it later.

However, I am not very clear about what should Eventmesh do with schema.

As far as I am concerned, there are four main concepts of schema: contract, serialization, deserialization, and validity.
- For contract, a schema is a contracted format of message between producer and consumer.
- For serialization, a producer serializes a message according to the contracted schema.
- For deserialization, a consumer deserializes a message according to the contracted schema.
- For validity, usually, a consumer checks the validity of the inbound serialized message before deserializing it.

So I was thinking that Eventmesh could take care of validity. If the inbound message from the producer is not obeying the specified schema, Eventmesh will discard this message and send an error status to the producer. The corresponding modules could be ```eventmesh-schema-json ```, ```eventmesh-schema-avro ```, and ```eventmesh-schema-protobuf ```.

Also, I get your point. Thinking from object-oriented programming, schema plugins could also provide the functions of ```serialize``` and ```deserialize```. Correct me if I had a misunderstanding. 


What do you think?

**Comment by ruanwenjun** [comment]
> 

You are right, the api of schema plugin should provide the 4 interface `contract, serialization, deserialization, and validity `.
I think it shouldn't take care about how to store the schema(like register schema).

**Comment by jzhou59** [comment]
> > 
> 
> Yes, you are right, the api of schema plugin should provide the 4 interface `contract, serialization, deserialization, and validity `. I think it shouldn't take care about how to store the schema(like register schema).

Ok, I have removed the schema storage-related functions from the schema plugin. What do you think of the current schema plugin structure of design?

Next, I will implement three examples of schema: JSON, Avro, and Protobuf respectively.

Also, where do you think the schema storage-related functions should be put? Maybe newly create a ```eventmesh-schemaregistry-plugin```?

**Comment by ruanwenjun** [comment]
> 

If the schema should be stored, it can be seen as a metadata for topic, so can be implemented i metadata store module.

**Comment by jzhou59** [comment]
Oh, not only storage, I mean the interaction between Eventmesh and openschemaregistry(such as retrieving, caching, and registering). Where should they be put?

**Comment by ruanwenjun** [comment]
> Oh, not only storage, I mean the interaction between Eventmesh and schemaregistry(such as retrieving, caching, and registering). Here schemaregistry is different from schema. Since schema-plugin is only designed as schema-related. Where should schemaregistry be put?

The schema is seen as the metadata of a topic. When you register a schema to a topic, you just bind the schema to the topic, so you don't need to add a `SchemaRegistey`, we will have many methods to modify the metadata, we will a Metadata module to manage these data.

**Pull request closed without merging**